### PR TITLE
[BB2-896] Responsive Design NodeJS React Sample

### PR DESCRIPTION
**Pull request by JFU-GIT**
**JIRA Ticket:**
[BB2-896](https://jira.cms.gov/browse/BB2-896)

**User Story or Bug Summary:**

As a 3rd party app, I should be able to understand what to do when I receive certain error messages from BB2 so that I can properly respond to my own users

AC:

Draft documentation for review for:

- What to do when receiving a 403 error
- What to do when receiving a 404 error
- What to do when receiving a 500 error
- how to implement exponential backoff

In the Node.JS and Python sample apps add in responsive design to show:

- What to do when receiving a 403 error
- What to do when receiving a 404 error
- What to do when receiving a 500 error
- how to implement exponential backoff

### What Does This PR Do?

1. Create a document [BB2 API Error Responses & Handling](https://confluence.cms.gov/pages/viewpage.action?pageId=540087166) where Blue Button 2.0 API Error Responses (HTTP status codes) and Recommended Error Handling Are Documented, including, among other things, how to retry request upon certain error responses.
2. Add improved error reporting to the sample app where it is applicable, and implement retry logic (exponential back off) when applicable.

Extra:

Avoid tweaking of the code for native OS run or in Docker run, introduce a shell script to start the sample in native OS or in Docker, make client and server code context aware such that the AUTH2 authentication flow functions when the sample is running in docker or on native OS.

### What Should Reviewers Watch For?

* Are error responses from API calls handled as instructed by the document?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is the retry work as expected?

If you're reviewing this PR, please check these things, in particular:

* Review the document on API error responses
* Verify the retry logic works (exponential back off)

To facilitate the verification of retry failed requests, an instrumented BB2 server is deployed to sandbox canary:
put below line into your host file on you local machine:

10.246.44.56	sandbox.bluebutton.cms.gov
(new canary deployed on Jan 19, 2022)

then:

1. create an app on sandbox (or using an existing one)
2. make sure the url is among the redirect urls: http://server:3001/api/bluebutton/callback/
3. copy the client_id, client_secret to the sample app server config.ts
4. start sample app in docker: go to local repo base and run ./start.sh docker

Wait until the server and client components are up and running.
note, a waffle switch "simulate_500_error" is provisioned on sandbox which when "active", will produce a 500 error on fhir request, with error message: "Service temporarily unavailable, try again later, network connection refused".

From client webUI, click button "Authorize" to complete the beneficiary medicare.gov login, and grant access to the app, the corresponding EOB search will hit the simulated 500 network connection error, and the failed EOB request met the criteria for retry.

below is a sample server stdout showing the retry process, the 3rd attempt succeeded (by setting switch "simulate_500_error" to false soon after 2nd retry ...)

`server_1  | [2021-11-19T00:16:56.149Z] INFO: [object Object]
server_1  | Error message: [ Request failed with status code 500 ]
server_1  | response code: 500
server_1  | response text: [object Object]
server_1  | Request failed and is retryable, entering retry process...
server_1  | wait  5  seconds...
server_1  | retry attempts:  1
server_1  | retry error: [ Request failed with status code 500 ]
server_1  | response code:  500
server_1  | response data:  {
server_1  |   detail: 'Service temporarily unavailable, try again later, network connection refused.
'
server_1  | }
server_1  | wait  10  seconds...
server_1  | retry attempts:  2
server_1  | retry error: [ Request failed with status code 500 ]
server_1  | response code:  500
server_1  | response data:  {
server_1  |   detail: 'Service temporarily unavailable, try again later, network connection refused.
'
server_1  | }
server_1  | wait  20  seconds...
server_1  | retry attempts:  3
server_1  | retry successful:
server_1  | {
server_1  |   resourceType: 'Bundle',
server_1  |   id: '787094eb-dca9-4af4-9b39-87e040f65417',
server_1  |   meta: { lastUpdated: '2021-09-23T13:36:42.530-04:00' },
server_1  |   type: 'searchset',
server_1  |   total: 51,
server_1  |   link: [
server_1  |     {
server_1  |       relation: 'first',
server_1  |       url: 'https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit?_format=appl
ication%2Fjson%2Bfhir&startIndex=0&_count=10&patient=-20000000000002'
server_1  |     },
server_1  |     {
server_1  |       relation: 'next',
server_1  |       url: 'https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit?_format=appl
ication%2Fjson%2Bfhir&startIndex=10&_count=10&patient=-20000000000002'
server_1  |     },
server_1  |     {
server_1  |       relation: 'last',
server_1  |       url: 'https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit?_format=appl
ication%2Fjson%2Bfhir&startIndex=50&_count=10&patient=-20000000000002'
server_1  |     },
server_1  |     {
server_1  |       relation: 'self',
server_1  |       url: 'https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit/?_count=10&_
format=application%2Fjson%2Bfhir&patient=-20000000000002&startIndex=0'
server_1  |     }
server_1  |   ],
server_1  |   entry: [
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] }
server_1  |   ]
server_1  | }
server_1  | config:
server_1  | {
server_1  |   url: 'https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit/',
server_1  |   method: 'get',
server_1  |   headers: {
server_1  |     Accept: 'application/json, text/plain, */*',
server_1  |     Authorization: 'Bearer k8O6XDxBf2RCLmD0xzAiGKlWvnOcRS',
server_1  |     'User-Agent': 'axios/0.21.1'
server_1  |   },
server_1  |   params: {
server_1  |     code: 'mRpSQkterTO9xzIfXEab6hRCun4ka0',
server_1  |     state: 'k6hb1zQg_TnBOsmozJVEyDwFxxLHd2tJrD8Q39WPjzE'
server_1  |   },
server_1  |   transformRequest: [ [Function: transformRequest] ],
server_1  |   transformResponse: [ [Function: transformResponse] ],
server_1  |   timeout: 0,
server_1  |   adapter: [Function: httpAdapter],
server_1  |   xsrfCookieName: 'XSRF-TOKEN',
server_1  |   xsrfHeaderName: 'X-XSRF-TOKEN',
server_1  |   maxContentLength: -1,
server_1  |   maxBodyLength: -1,
server_1  |   validateStatus: [Function: validateStatus],
server_1  |   data: undefined
server_1  | }
`

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies? **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following pre-requisite changes have been fully deployed:

* CMSgov/some_repo#42
  
### Any Migrations?

* [ ] Yes, there are migrations
  * [ ] The migrations should be run PRIOR to the code being deployed
  * [ ] The migrations should be run AFTER the code is deployed
  * [ ] There is a more complicated migration plan (downtime, etc) 
* [x] No migrations

### Submitter Checklist

I have gone through and verified that...:

* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.
… docker or natively.

### CODE REVIEW

**Comment by nbragdon** [comment]
@JFU-GIT We really shouldn't be creating new startup scripts. I think the idea of using docker was that it was platform agnostic. Now adding in the shell scripts makes it less so. I think we are just going to have to work with what is there, or somehow integrate it into the docker setup. I think this PR shouldn't change any of that though.

**Comment by JFU-GIT** [comment]
> @JFU-GIT We really shouldn't be creating new startup scripts. I think the idea of using docker was that it was platform agnostic. Now adding in the shell scripts makes it less so. I think we are just going to have to work with what is there, or somehow integrate it into the docker setup. I think this PR shouldn't change any of that though.

understand the point, but sometime start script reduce the manual steps...

File: server/src/utils/bb2.ts [hunk 1]
```diff
@@ -33,7 +33,8 @@ export async function getAccessToken(code: string, state: string | undefined) {
     const envConfig = config[db.settings.env];
 
     const BB2_ACCESS_TOKEN_URL = envConfig.bb2BaseUrl + '/' + db.settings.version + '/o/token/';
-
+    // const BB2_ACCESS_TOKEN_URL = 'http://host.docker.internal:8000/' + db.settings.version + '/o/token/';
```
Comment (nbragdon) [inline server/src/utils/bb2.ts:36]: Lets remove this commented out line

File: server/src/utils/bb2.ts [hunk 2]
```diff
@@ -1,9 +1,9 @@
-import axios from 'axios';
 import FormData from 'form-data';
 
 import db from './db';
 import config from '../configs/config';
 import { generateCodeChallenge, generateRandomState } from './generatePKCE';
+import { endpointPost } from './call'
```
Comment (nbragdon) [inline server/src/utils/bb2.ts:6]: Can we rename this file to something more descriptive than `call`?

**Review by nbragdon — COMMENTED** [review]
@JFU-GIT I don't think this cron/queue approach is really ideal for exponential backoff... I think we should do something more like in this article: https://advancedweb.hu/how-to-implement-an-exponential-backoff-retry-strategy-in-javascript/

Specifically the solution listed under: Rejection-based retrying for async/await syntax

File: server/src/utils/call.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+export async function endpointPost(endpoint_url: string, data: FormData, extra: any) {
+    return await httpCall({ 
+        method: 'post',
+        url: endpoint_url,
+        data: data,
+        headers: extra}, true);
+}
+
+export async function endpointGet(endpoint_url: string, req_qry: any, token: string) {
```
Comment (nbragdon) [inline server/src/utils/call.ts:14]: I'm thinking something more like... `getFhirResource` might be a better naming schema here

Comment (nbragdon) [reply server/src/utils/call.ts:14]: I think we only would want to do retries when requesting fhir data, I don't think we want to introduce them into the auth flow

File: README.md [hunk 4]
```diff
@@ -57,3 +57,33 @@ Read the DEVELOPER NOTES found in the code to understand the application
 and where you will need to make adjustments/changes as well as some 
 suggestions for best practices.
 
+Usage Examples
```
Comment (nbragdon) [inline README.md:60]: Lets keep the startup instructions as they were

File: docker-compose.yml [hunk 5]
```diff
@@ -7,6 +7,7 @@ services:
       dockerfile: ./Dockerfile
     ports: 
       - "3001:3001"
+      - "9229:9229"
```
Comment (nbragdon) [inline docker-compose.yml:10]: why was the new port added?

Comment (JFU-GIT) [reply docker-compose.yml:10]: for remote debugging

File: server/src/utils/call.ts (hunk 3 above)
Comment (JFU-GIT) [reply server/src/utils/call.ts:15]: it's a unified code so we have a centralized place to add control, in case something else needed, it can be added here conveniently.

File: server/src/index.ts [hunk 6]
```diff
@@ -1,7 +1,11 @@
 import './pre-start'; // Must be the first import
 import app from '@server';
 import logger from '@shared/Logger';
+import { schedule } from './utils/retry';
 
+// start retry cron job
+logger.info('Schedule retry cron job...');
+schedule();
```
Comment (dtisza1) [inline server/src/index.ts:8]: > @JFU-GIT I don't think this cron/queue approach is really ideal for exponential backoff... I think we should do something more like in this article: https://advancedweb.hu/how-to-implement-an-exponential-backoff-retry-strategy-in-javascript/
> 
> Specifically the solution listed under: Rejection-based retrying for async/await syntax

@JFU-GIT @nbragdon I agree that the use of node-cron scheduler and queuing seems complicated for the purpose.

I just read through the article and agree that the "rejection-based retrying" method seems like it is simpler and a more typical/standard way to implement this.  @nbragdon Have you used this type of method successfully in the past?

**Review by dtisza1 — CHANGES_REQUESTED** [review]
@JFU-GIT Just a few review comments. TY!

File: start.sh [hunk 7]
```diff
@@ -0,0 +1,95 @@
+#!/bin/bash
+
+#
+# Start the sample application:
+# 1) in docker or
+# 2) in native OS - type CTRL C to terminate both the client and server components of the sample app
+#
```
[the remark this answers is not in the document]
Comment (JFU-GIT) [reply start.sh:7]: no problem to remove the start.sh script, the sample app is a node js react app and developer playing with it mostly know how to fire it up either in docker or natively, just add help info in readme seems sufficient.

[the remark this answers is not in the document]
Comment (dtisza1) [reply start.sh:7]: @JFU-GIT That sounds great and the added help info would be nice to have.

File: server/src/index.ts (hunk 6 above)
Comment (JFU-GIT) [reply server/src/index.ts]: a in place retry added.

File: client/package.json [hunk 8]
```diff
@@ -23,7 +23,8 @@
     "web-vitals": "^1.0.1"
   },
   "scripts": {
-    "start": "react-scripts start",
+    "start": "REACT_APP_CTX=docker react-scripts start",
+    "start2": "REACT_APP_CTX=native react-scripts start",
```
Comment (nbragdon) [inline client/package.json:27]: Let's rename this to something more descriptive. `start2` isn't really telling me how/why it's different. Something like `start-native` would be more descriptive.

File: client/src/setupProxy.js [hunk 9]
```diff
@@ -4,7 +4,7 @@ module.exports = function(app) {
     app.use(
         '/api',
         createProxyMiddleware({
-            target: 'http://server:3001',
+            target: (process.env.REACT_APP_CTX === 'docker' ? 'http://server:3001' : 'http://localhost:3001'),
```
Comment (nbragdon) [inline client/src/setupProxy.js:7]: Nice change 👍

File: client/yarn.lock [hunk 10]
```diff
@@ -3224,9 +3224,9 @@ caniuse-api@^3.0.0:
     lodash.uniq "^4.5.0"
 
 caniuse-lite@^1.0.0, caniuse-lite@^1.0.30000981, caniuse-lite@^1.0.30001109, caniuse-lite@^1.0.30001125, caniuse-lite@^1.0.30001181:
-  version "1.0.30001191"
-  resolved "https://registry.yarnpkg.com/caniuse-lite/-/caniuse-lite-1.0.30001191.tgz#bacb432b6701f690c8c5f7c680166b9a9f0843d9"
-  integrity sha512-xJJqzyd+7GCJXkcoBiQ1GuxEiOBCLQ0aVW9HMekifZsAVGdj5eJ4mFB9fEhSHipq9IOk/QXFJUiIr9lZT+EsGw==
+  version "1.0.30001279"
+  resolved "https://registry.npmjs.org/caniuse-lite/-/caniuse-lite-1.0.30001279.tgz"
```
Comment (nbragdon) [inline client/yarn.lock:3228]: Looks like the registry changed from yarn to npm... any idea why this change was made?

File: server/package.json [hunk 11]
```diff
@@ -70,9 +70,11 @@
     "jet-logger": "^1.0.4",
     "jsonfile": "^6.1.0",
     "module-alias": "^2.2.2",
-    "morgan": "^1.10.0"
+    "morgan": "^1.10.0",
+    "node-cron": "^3.0.0"
```
Comment (nbragdon) [inline server/package.json:74]: Let's remove this package since it's no longer used

File: server/package.json [hunk 12]
```diff
@@ -70,9 +70,11 @@
...
   },
   "devDependencies": {
+    "@types/node-cron": "^3.0.0",
```
Comment (nbragdon) [inline server/package.json:77]: Can also remove the typing

File: server/src/routes/Authorize.ts [hunk 13]
```diff
@@ -52,6 +52,7 @@ export async function authorizationCallback(req: Request, res: Response) {
         * to display or store the error
          */
         console.log(e);
+        logger.info(e)
```
Comment (nbragdon) [inline server/src/routes/Authorize.ts:55]: If we want to start using the logger, we should probably move away from the console logs. Also in this case it should be `logger.error(e)`.

File: server/src/routes/Data.ts [hunk 14]
```diff
@@ -15,15 +15,9 @@ import { getLoggedInUser } from 'src/utils/user';
...
-    
-    const response = await axios.get(BB2_BENEFIT_URL, { 
-        params: req.query,
-        headers: {
-            'Authorization': `Bearer ${loggedInUser.authToken?.access_token}`
-        }
-    });
-    return response.data;
+    // get EOB end point
+    const response = await get(envConfig.bb2BaseUrl + '/' + db.settings.version + '/fhir/ExplanationOfBenefit/', req.query, `${loggedInUser.authToken?.access_token}`);
```
Comment (nbragdon) [inline server/src/routes/Data.ts:19]:
```suggestion
    const response = await get(`${envConfig.bb2BaseUrl}/${db.settings.version}/fhir/ExplanationOfBenefit/`, req.query, `${loggedInUser.authToken?.access_token}`);
```

File: server/src/routes/Data.ts [hunk 15]
```diff
@@ -42,45 +36,24 @@ export async function getBenefitDataEndPoint(req: Request, res: Response) {
...
-    
-    const response = await axios.get(BB2_PATIENT_URL, { 
-        params: req.query,
-        headers: {
-            'Authorization': `Bearer ${loggedInUser.authToken?.access_token}`
-        }
-    });
-
+    // get Patient end point
+    const response = await get(envConfig.bb2BaseUrl + '/' + db.settings.version + '/fhir/Patient/', req.query, `${loggedInUser.authToken?.access_token}`);
```
Comment (nbragdon) [inline server/src/routes/Data.ts:40]:
```suggestion
    const response = await get(`${envConfig.bb2BaseUrl}/${db.settings.version}/fhir/Patient/`, req.query, `${loggedInUser.authToken?.access_token}`);
```

File: server/src/routes/Data.ts [hunk 16]
```diff
@@ -42,45 +36,24 @@ export async function getBenefitDataEndPoint(req: Request, res: Response) {
...
-    
-    const response = await axios.get(BB2_COVERAGE_URL, { 
-        params: req.query,
-        headers: {
-            'Authorization': `Bearer ${loggedInUser.authToken?.access_token}`
-        }
-    });
-
+    // get Coverage end point
+    const response = await get(envConfig.bb2BaseUrl + '/' + db.settings.version + '/fhir/Coverage/', req.query, `${loggedInUser.authToken?.access_token}`);
```
Comment (nbragdon) [inline server/src/routes/Data.ts:48]:
```suggestion
    const response = await get(`${envConfig.bb2BaseUrl}/${db.settings.version}/fhir/Coverage/`, req.query, `${loggedInUser.authToken?.access_token}`);
```

File: server/src/routes/Data.ts [hunk 17]
```diff
@@ -42,45 +36,24 @@ export async function getBenefitDataEndPoint(req: Request, res: Response) {
...
-    
-    const response = await axios.get(BB2_BENEFIT_URL, { 
-        params: req.query,
-        headers: {
-            'Authorization': `Bearer ${loggedInUser.authToken?.access_token}`
-        }
-    });
-
+    // get usrinfo end point
+    const response = await get(envConfig.bb2BaseUrl + '/' + db.settings.version + '/connect/userinfo', req.query, `${loggedInUser.authToken?.access_token}`);
```
Comment (nbragdon) [inline server/src/routes/Data.ts:56]:
```suggestion
    const response = await get(`${envConfig.bb2BaseUrl}/${db.settings.version}/connect/userinfo`, req.query, `${loggedInUser.authToken?.access_token}`);
```

File: server/src/utils/request.ts [hunk 18]
```diff
@@ -0,0 +1,101 @@
+import axios from 'axios';
+import FormData from 'form-data';
+import logger from '@shared/Logger';
+
+export async function post(endpoint_url: string, data: FormData, extra: any) {
```
Comment (nbragdon) [inline server/src/utils/request.ts:5]:
```suggestion
export async function post(endpoint_url: string, data: FormData, headers: any) {

```
We should avoid using the `any` typing whenever possible.. not sure if the headers have a generic typing or not. If so you should be able to get it from the response type of this function: `form.getHeaders()`

File: server/src/utils/request.ts [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+
+export async function post(endpoint_url: string, data: FormData, extra: any) {
+    return await request({ 
+        method: 'post',
+        url: endpoint_url,
+        data: data,
+        headers: extra}, true);
+}
+
+export async function get(endpoint_url: string, req_qry: any, token: string) {
```
Comment (nbragdon) [inline server/src/utils/request.ts:13]:
```suggestion
export async function get(endpointUrl: string, params: any, authToken: string) {

```

File: server/src/utils/request.ts [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+    return await request({ 
+        method: 'get',
+        url: endpoint_url,
+        params: req_qry,
+        headers: {
+            'Authorization': `Bearer ${token}`
+        }}, true);
+}
+
+export async function request(config: any, retry_flag: boolean) {
```
Comment (nbragdon) [inline server/src/utils/request.ts:23]:
```suggestion
export async function request(config: any, retryFlag: boolean) {

```

File: server/src/utils/request.ts [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+            console.log("error.request: " + error.request);
+        }
+        // dump axios config for diagnosis
+        console.log("config:")
+        console.log(error.config);
+    }
+    return resp    
+}
+
+function is_retryable(error: any) {
```
Comment (nbragdon) [inline server/src/utils/request.ts:60]:
```suggestion
function isRetryable(error: Error) {

```

Comment (nbragdon) [reply server/src/utils/request.ts:60]: The convention for JS is to use camel case, not snake case. Kind of weird to have mixed cases throughout the code base.

File: server/src/utils/request.ts [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+        headers: {
+            'Authorization': `Bearer ${token}`
+        }}, true);
+}
+
+export async function request(config: any, retry_flag: boolean) {
+    var resp = null
+    try {
+        resp = await axios(config);
+    } catch (error: any) {
```
Comment (nbragdon) [inline server/src/utils/request.ts:27]:
```suggestion
    } catch (error: Error) {

```

File: server/src/utils/request.ts [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+        method: 'get',
+        url: endpoint_url,
+        params: req_qry,
+        headers: {
+            'Authorization': `Bearer ${token}`
+        }}, true);
+}
+
+export async function request(config: any, retry_flag: boolean) {
+    var resp = null
```
Comment (nbragdon) [inline server/src/utils/request.ts:24]:
```suggestion
    let resp = null

```
In modern JS it's common to use either `let` if the variable will change over time or `const` if it will never change.

File: server/src/utils/request.ts [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+    const max_attempts = 3
+    var resp = null
+    for (let i = 0; i < max_attempts; i++) {
+        var wait_in_sec = interval * (2 ** i)
+        console.log("wait ", wait_in_sec, " seconds...")
+        await sleep(wait_in_sec * 1000)
+        console.log("retry attempts: ", i+1)
+        try {
+            resp = await axios(config);
+            console.log("retry successful:")
```
Comment (nbragdon) [inline server/src/utils/request.ts:82]: Can we clean up some of these console logs?

File: server/src/utils/request.ts [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+            if (error.response) {
+                console.log("response code: ", error.response.status)
+                console.log("response data: ", error.response.data)
+            }
+        }
+    }
+    return resp
+}
+
+function sleep(ms: any) {
```
Comment (nbragdon) [inline server/src/utils/request.ts:96]:
```suggestion
function sleep(time: number) {

```

File: ErrorResponses.md [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+## Overview
+
+This document serves as a supplementary to Blue Button 2.0 API Developer Documents, it gives more details on most commonly error responses and recommended error handling on the client side, e.g. guide lines on retrying the requests on certain errors.
+
+## Error Responses and Client Actions
+
+### Authorization Requests
+
+
+| Status Code    | End Point URL   | Error Message              | Action            | Comments                                               |
```
Comment (nbragdon) [inline ErrorResponses.md:12]: I'm trying to decide if this is just too much information. Should we just highlight which type of requests they could retry on? Everything else feels a bit too much for a sample app. Basically it's a lot of text that essentially sums up to: "Don't do anything".

Comment (JFU-GIT) [reply ErrorResponses.md:12]: though the purpose is to give more bb2 specifics, since HTTP response codes are already defined.

Can reduce the verbose-ness a bit.

File: server/src/utils/request.ts (hunk 24 above)
Comment (JFU-GIT) [reply server/src/utils/request.ts:82]: those are there to help reviewers verify the retry is working..., otherwise, it'd be more involved - e.g. do source line debugging...

File: client/yarn.lock (hunk 10 above)
Comment (JFU-GIT) [reply client/yarn.lock:3228]: probably used below to update the browser db up to date to silence the warnings:
npx browserslist@latest --update-db

not sure if yarn upgrade works?

yarn upgrade caniuse-lite browserlist

File: ErrorResponses.md [hunk 27]
```diff
@@ -0,0 +1,41 @@
+# Blue Button 2.0 API Error Responses
+
+## Overview
+
+This document serves as a supplementary to Blue Button 2.0 API Developer Documents, it gives more details on most commonly error responses and recommended error handling on the client side, e.g. guide lines on retrying the requests on certain errors.
```
Comment (nbragdon) [inline ErrorResponses.md:5]:
```suggestion
This document serves as a supplementary to Blue Button 2.0 API Developer Documents. It gives more details on the most common error responses and how to properly handle them.

```

File: ErrorResponses.md [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+
+
+### Retry
+
+App can choose to retry the requests on certain 500 errors heuristically depending on the root cause as indicated by the error messages, e.g. backend server is unavailable due to temporarily network down etc., the app can choose from various retry patterns that best fits the use case (e.g. exponential back off).
+
+Auto retrying on FHIR Data read/search to overcome a FHIR backend network temporary downtime is recommended, but with sensible retry settings.
+
+Due to the involvement of end user (beneficiary), auto retrying on requests in an authorization flow does not seem practical, hence not recommended.
```
Comment (nbragdon) [inline ErrorResponses.md:41]:
```suggestion
Due to the involvement of the end user (beneficiary), auto retrying requests in the authorization flow are not recommended.

```

File: ErrorResponses.md [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+
+
+### Retry
+
+App can choose to retry the requests on certain 500 errors heuristically depending on the root cause as indicated by the error messages, e.g. backend server is unavailable due to temporarily network down etc., the app can choose from various retry patterns that best fits the use case (e.g. exponential back off).
+
+Auto retrying on FHIR Data read/search to overcome a FHIR backend network temporary downtime is recommended, but with sensible retry settings.
```
Comment (nbragdon) [inline ErrorResponses.md:39]:
```suggestion
Auto retrying (with sensible retry settings) on FHIR Data read/search to overcome a FHIR backend network temporary downtime is recommended.

```

File: ErrorResponses.md [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+| 500<br>SERVER ERROR | /v[12]/fhir/.* | The root cause of the 500 error, some times, is indicated by the error message, <br>the app can choose to retry the failed request depend on the nature of the root cause, <br>examples that might be retriable are those related to network down (temporarily):<br>Example:<br>ConnectionError at /mymedicare/login<br>HTTPSConnectionPool(host='test.accounts.cms.gov', port=443): <br>Max retries exceeded with url: /health (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f46599dafd0>: <br>Failed to establish a new connection: [Errno -2] Name or service not known')) | Heuristic on Retry | App can choose to retry on some of the 500 errors as shown by the example, this is a heuristic approach. |
+
+
+### Retry
+
+App can choose to retry the requests on certain 500 errors heuristically depending on the root cause as indicated by the error messages, e.g. backend server is unavailable due to temporarily network down etc., the app can choose from various retry patterns that best fits the use case (e.g. exponential back off).
```
Comment (nbragdon) [inline ErrorResponses.md:37]:
```suggestion


```

File: ErrorResponses.md (hunk 26 above)
Comment (nbragdon) [reply ErrorResponses.md:13]: Want to take an attempt at trimming it down? Really focus on what they would need to know in order to successfully build an application against BB2.

Comment (JFU-GIT) [reply ErrorResponses.md:13]: those are frequently used error responses among the full set http status codes, and it contains bb2 specific info, so there are added values, or let me know what should be trimmed? don't mind remove the whole mark down - remember it was added based on some feedback-:

**Comment by nbragdon** [comment]
@JFU-GIT Reminder that this also has to be done for the python app as well

File: ErrorResponses.md (hunk 26 above)
Comment (nbragdon) [reply ErrorResponses.md:13]: My only concern is that this table is a lot of text. I feel like most people wont read it all, and then they are missing out on important information. Were we to trim it down to just when they need to do something (ie retry) then they might catch it. If you feel strongly we can leave it though. Just clean up the typing issues I pointed out and we should be able to get this approved 👍

Comment (JFU-GIT) [reply ErrorResponses.md:13]: for the enforce of typing, for those places that the type of the var or param is obvious, explicit type is given, but some places e.g. the err var of axios try catch, a single type could not cover all situations, e.g. the error thrown from axios end point call could be AxiosError but it could be something else, there are discussion on this from internet..., also ran across some people's code to handle the multiple error types - but it will pull in more code, may be we could log a backlog for "making node js react sample strong typed" as future improvement? thanks for brought this up.

File: client/src/components/records.tsx [hunk 31]
```diff
@@ -31,49 +37,79 @@ export default function Records({ }) {
...
-                    const resource = resourceData.resource;
-                    return {
-                        id: resource.id,
-                        code: resource.item[0]?.productOrService?.coding[0]?.code || 'Unknown',
-                        display: resource.item[0]?.productOrService?.coding[0]?.display || 'Unknown Prescription Drug',
-                        amount: resource.item[0]?.adjudication[7]?.amount?.value || '0'
-                    }
-                });
-                setRecords(records);
+                if (eobData.hasOwnProperty('entry')) {
```
Comment (nbragdon) [inline client/src/components/records.tsx:40]:
```suggestion
                if (eobData.entry) {
```

File: server/src/utils/request.ts (hunk 18 above)
Comment (nbragdon) [reply server/src/utils/request.ts:4]: Can you switch from snake -> camel?

**Comment by JFU-GIT** [comment]
@nbragdon @dtisza1 rebased and retested plus some final touches on error handling - server logs error with code + details, while client got generic message on failed auth or failed fhir query.

File: server/src/utils/request.ts (hunk 24 above)
Comment (dtisza1) [reply server/src/utils/request.ts:85]: @JFU-GIT Should the use of `console.log()` be converted to use the '@shared/Logger' instead here? And other locations for consistency across the code base?

**Review by dtisza1 — CHANGES_REQUESTED** [review]
@JFU-GIT  This is looking good to me! Just having some possible changes and comments below:

* ErrorResponses.md:  This looks good in table format and split out from the Readme.md.  I think it would be good to put a link in the Readme.md that references this info.
* Readme.md:  Looks good to me. TY for adding the native startup! Just noticed it could use some cleanup with some sentences. For example, some sentences could begin capitalized.
* Should the console type logging be replaced in all areas with the '@shared/Logger'?
* Should the code be edited to use Camel case? Especially for the new code added in this PR. Could address previous code separately.
* The retry logic looks good to me! I didn't test it, but from your testing notes and server logs it looks to work as expected.

Thank you!

**Comment by JFU-GIT** [comment]
@nbragdon @dtisza1 addressed feedbacks in the last commit, also deployed a sbx canary with instrumented BB2:

put below in hosts if want to test retry:

10.246.44.56 sandbox.bluebutton.cms.gov

**Review by dtisza1 — APPROVED** [review]
This looks good to me!

TY for the updates on the change requests.

I used the canary you set up in Sandbox for testing and everything worked as expected with the retries.

Good work on this!

File: server/eslintrc.json [hunk 32]
```diff
@@ -0,0 +1,20 @@
+{
+    "env": {
+        "browser": true,
```
Comment (nbragdon) [inline server/eslintrc.json:3]: I think for the server, `browser` shouldn't be true?

File: server/package.json [hunk 33]
```diff
@@ -5,8 +5,9 @@
     "build": "./node_modules/.bin/ts-node build.ts",
     "lint": "eslint . --ext .ts",
     "start": "node -r module-alias/register ./dist --env=production",
+    "start:debug": "node --inspect=0.0.0.0:9229 ./node_modules/.bin/ts-node -r tsconfig-paths/register ./src",
     "start:dev": "nodemon",
-    "test": "nodemon --config ./spec/nodemon.json"
+    "test": "mocha --exit -r ts-node/register spec/*.spec.ts"
```
Comment (nbragdon) [inline server/package.json:10]: Looks like some testing changes might have snuck through to this PR

**Review by nbragdon — APPROVED** [review]
Nice job on this, thanks for making the requested changes!
